galley/pkg/mcp/snapshot/snapshot_test.go

```diff
@@ -0,0 +1,327 @@
+func (f fakeTypeBase) String() string           { return f.Info }
+func (f fakeTypeBase) ProtoMessage()            {}
+func (f fakeTypeBase) Marshal() ([]byte, error) { return []byte(f.Info), nil }
+
+type fakeType0 struct{ fakeTypeBase }
+type fakeType1 struct{ fakeTypeBase }
+type fakeType2 struct{ fakeTypeBase }
+
+const (
+	typePrefix      = "type.googleapis.com/"
+	fakeType0Prefix = "istio.io.galley.pkg.mcp.server.fakeType0"
+	fakeType1Prefix = "istio.io.galley.pkg.mcp.server.fakeType1"
+	fakeType2Prefix = "istio.io.galley.pkg.mcp.server.fakeType2"
+
+	fakeType0TypeURL = typePrefix + fakeType0Prefix
+	fakeType1TypeURL = typePrefix + fakeType1Prefix
+	fakeType2TypeURL = typePrefix + fakeType2Prefix
+)
+
+func mustMarshalAny(pb proto.Message) *types.Any {
+	a, err := types.MarshalAny(pb)
+	if err != nil {
+		panic(err.Error())
+	}
+	return a
+}
+
+func init() {
+	proto.RegisterType((*fakeType0)(nil), fakeType0Prefix)
+	proto.RegisterType((*fakeType1)(nil), fakeType1Prefix)
+	proto.RegisterType((*fakeType2)(nil), fakeType2Prefix)
+
+	fakeResource0 = &mcp.Envelope{
+		Metadata: &mcp.Metadata{Name: "f0"},
+		Resource: mustMarshalAny(&fakeType0{fakeTypeBase{"f0"}}),
+	}
+	fakeResource1 = &mcp.Envelope{
+		Metadata: &mcp.Metadata{Name: "f1"},
+		Resource: mustMarshalAny(&fakeType1{fakeTypeBase{"f1"}}),
+	}
+	fakeResource2 = &mcp.Envelope{
+		Metadata: &mcp.Metadata{Name: "f2"},
+		Resource: mustMarshalAny(&fakeType2{fakeTypeBase{"f2"}}),
+	}
+}
+
+type fakeSnapshot struct {
+	// read-only fields - no locking required
+	resources map[string][]*mcp.Envelope
+	versions  map[string]string
+}
+
+func (fs *fakeSnapshot) Resources(typ string) []*mcp.Envelope { return fs.resources[typ] }
+func (fs *fakeSnapshot) Version(typ string) string            { return fs.versions[typ] }
+
+func (fs *fakeSnapshot) copy() *fakeSnapshot {
+	fsCopy := &fakeSnapshot{
+		resources: make(map[string][]*mcp.Envelope),
+		versions:  make(map[string]string),
+	}
+	for typeURL, resources := range fs.resources {
+		fsCopy.resources[typeURL] = append(fsCopy.resources[typeURL], resources...)
+		fsCopy.versions[typeURL] = fs.versions[typeURL]
+	}
+	return fsCopy
+}
+
+func makeSnapshot(version string) *fakeSnapshot {
+	return &fakeSnapshot{
+		resources: map[string][]*mcp.Envelope{
+			fakeType0TypeURL: {fakeResource0},
+			fakeType1TypeURL: {fakeResource1},
+			fakeType2TypeURL: {fakeResource2},
+		},
+		versions: map[string]string{
+			fakeType0TypeURL: version,
+			fakeType1TypeURL: version,
+			fakeType2TypeURL: version,
+		},
+	}
+}
+
+var _ Snapshot = &fakeSnapshot{}
+
+var (
+	node = &core.Node{
+		Id:      "test-id",
+		Cluster: "test-cluster",
+	}
+
+	key = node.Id
+
+	fakeResource0 *mcp.Envelope
+	fakeResource1 *mcp.Envelope
+	fakeResource2 *mcp.Envelope
+
+	WatchResponseTypes = []string{
+		fakeType0TypeURL,
+		fakeType1TypeURL,
+		fakeType2TypeURL,
+	}
+)
+
+// TODO - refactor tests to not rely on sleeps
+var (
+	deferredResponseTimeout = 200 * time.Millisecond
```

Comment:
Given that cache is a purely inactive component, do you really need to check for delayed responses? It should be guaranteed that there will be no delayed response after the initial response, unless the snapshot gets updated.

Or am I missing something?

Reply:
The intent is to verify that we either receive an immediate response (via return) or async response (via channel) but not both. Negative test cases for async response wait `deferredResponseTimeout` to verify no response is received if a snapshot wasn't available.